### ISSUE TITLE
Bookmark labels not visible in Thorium3.2.0-alpha.1.13614884238

### ISSUE THREAD

**Issue by GeorgeKerscher** [edrlab/thorium-reader#2829]
Using NVDA, the bookmark labels are not visible. I find a location I want to bookmark and press NVDA key + enter to report the exact location. I then press CTRL + B to insert a bookmark. It is inserted properly and I can navigate to it from the bookmark list. However, when reviewing the list of bookmarks, the label, such as bookmark 1 is not shown. What is shown is a button to go to the bookmark with a percentage, a button to edit, and a dialog to delete.

I can edit the label, which will have bookmark 1 or 2, etc. I can delete that and put in my own meaningful text.

This was working in the past and I do not know in which beta this behavior changed. It has been in the last few beta builds.
@nvdaes

**Comment by nvdaes** [comment]
I confirm this regression. Previously, the multiline text set for bookmarks was presented.

**Comment by danielweck** [comment]
Thank you George, I will investigate.

By the way, how do you navigate to the bookmarks list?
With Voice Over on Mac I am forced to tab into the list, as opposed to just using CONTROL OPTION (the VO special key combination) plus left/right arrow to explore the GUI.

The tab GUI control has tab items that automatically show their contents when focussed, instead of waiting for the user enter or space key to activate the tab contents. I don't know if this is standard practice, Thorium is using a GUI library since version 3.

Daniel

**Comment by GeorgeKerscher** [comment]
Normally I press ctrl+n and arrow down to what I want. If I tab, I do not hear percentage and by pressing the down arrow, I do hear what is after the buttons.

I normally find the bookmark information after the list of navigation items.

If I select the TOC, it is expanded and it is unclear where that is. It seems to be placed at the end of the chapter. I think sometimes it appears within the navigation section. That is probably another issue and I wanted to get the Bookmark issue addressed now that we have NVDA + Enter to mark my location.

**Comment by nvdaes** [comment]
I think that if control+m could be used to open bookmarks, or control+shift+m, to be more similar to the keystroke for the TOC, it would be great.
And if a click event could be sent when pressing control+b. Since NVDA+enter is something discovered by us, not explicitly documented.

**Comment by danielweck** [comment]
> And if a click event could be sent when pressing control+b. Since NVDA+enter is something discovered by us, not explicitly documented.

Unless the mouse cursor is exactly on top of the element (paragraph, sentence, word) that the screen reader is currently speaking (i.e. the internal virtual buffer cursor) then Thorium forcing a click event is bound to cause problems (e.g. clicking on some completely unrelated GUI control, maybe even outside of Thorium). The alternative option is to generate a click event on the current reading location, but that is nonsensical because it assumes the user's reading location is already known, in which case CONTROL + B works already. I think you are referring to the screen reader internal speech cursor which Thorium cannot access as it is proprietary (there is no public API to fetch this information). So the screen reader user must somehow tell the HTML DOM in the webview that it is reading / listening at a particular location. My experience with Voice Over is that CONTROL + OPTION + SPACE works reliably, and if I understand you correctly NVDA + ENTER works too.

**Comment by danielweck** [comment]
> I think that if control+m could be used to open bookmarks, or control+shift+m, to be more similar to the keystroke for the TOC, it would be great.

At some point we will run out of keyboard shortcuts (smile), but yes this would definitely help. In this case I think Thorium should also offer hotkeys for annotations, goto-page, landmarks, etc. (why just TOC and bookmarks?)

**Comment by danielweck** [comment]
Fixed the bookmarks (and annotations) "go to the content" button label, now speaks the bookmark label and the annotation text too. Thorium is building: https://github.com/edrlab/thorium-reader/actions/runs/13630022469/

**Comment by nvdaes** [comment]
> So the screen reader user must somehow tell the HTML DOM in the webview that it is reading / listening at a particular location. My experience with Voice Over is that CONTROL + OPTION + SPACE works reliably, and if I understand you correctly NVDA + ENTER works too.

Yes, it works.
The problem is that this is not documented and sometimes may be forgoten to use NVDA+enter. I think that an NVDA-addon would be useful for this, or even an internal app module included in NVDA.
Do you agree? If for any reason you disagree, I won't develop it.

**Comment by nvdaes** [comment]
> At some point we will run out of keyboard shortcuts (smile), but yes this would definitely help. In this case I think Thorium should also offer hotkeys for annotations, goto-page, landmarks, etc. (why just TOC and bookmarks?)

Thorium offers control+shift+p, I think, for go to page. Annotations aren't currently accessible to screen reader users, so keystrokes, more useful for screen reader users, may not be needed for annotations. And landmarks seem not to be used in most books.

**Comment by GeorgeKerscher** [comment]
Works great!

**Comment by GeorgeKerscher** [comment]
I like the idea of an NVDA add-in for thorium that might do a number of things. It could be the same as the Thorium shortcut. Would NVDA keystroke take  precidence? It could be NVDA + CTRL + B
it would:
do a nvda + enter
issue ctrl + B
take you to that bookmark in edit mode
delete bookmark xxx
be ready for input of the name of the bookmark.
enter could save the bookmark and return to your bookmark location.

Thoughts?

**Comment by danielweck** [comment]
FYI, we are implementing the SHIFT CTRL B "create bookmark and immediately edit the label" feature.

**Comment by GeorgeKerscher** [comment]
Hi,

When will this be implemented. I am planning on doing a video for CSUN that demonstrates NVDA key + enter to mark a spot and then inserting the bookmark.

I could also demo the ctrl + shift + B if it is ready in time.

**Comment by danielweck** [comment]
> When will this be implemented. I am planning on doing a video for CSUN that demonstrates NVDA key + enter to mark a spot and then inserting the bookmark.
> 
> I could also demo the ctrl + shift + B if it is ready in time.

Hi George, CC @panaC 
We are going to put a high priority next week to ship this feature, it shouldn't be too hard. We already have the plumbing in place to directly jump into the bookmark and annotation sections of the navigation dialog (much like SHIFT-CTRL-N, but for a sighted user it works by clicking with the mouse cursor on document visual highlights or margin indicators).
Pierre, as you know when clicking an annotation the list item in the paginated resultst is automatically scrolled to and keyboard-focussed. We need this for bookmarks as well (from the click UX which just displays the user-clicked bookmark, but also in the upcoming SHIFT CONTROL B keyboard shortcut which will not only focus on the bookmark but also open the label editor ready for amending the default). We also need to fix the non-sensical bookmark numbering (when deleting "Bookmark 1", it seems to re-appear because the numbering is dynamic for the displayed GUI list, it is not intrinsic to the bookmark data model) Thank you!

**Comment by nvdaes** [comment]
@GeorgeKerscher , thanks for your plans with the demo about NVDA+enter. Hope you can share a link to see you.
cc: @nvaccess. Unfortunately, I think that this is not enough documented and probagly some people don't know how to create bookmarks with NVDA.
